party_menu/party_menu.lua

```diff
@@ -0,0 +1,521 @@
+    for i, row_window in ipairs(party_window_states) do
+        row_window.x = options.global_window_state.x + outer_padding
+        row_window.width = options.global_window_state.width - 2 * outer_padding
+        row_window.y = global_window_state.y + outer_padding * i + row_window_height * (i - 1)
+        row_window.height = row_window_height
+    end
+end
+
+local function get_row_string()
+    local name_job_hp_format_string = options.text_typeface .. ' ' .. tostring(options.text_size) .. 'px bold ' .. get_hp_color()
+    local name_job_hp = get_name() .. get_hp()
+    local name_job_hp_format = string.format('[' .. name_job_hp .. ']{' .. name_job_hp_format_string .. '}')
+
+    local mp_format_string = options.text_typeface .. ' ' .. tostring(options.text_size) .. 'px bold ' .. get_mp_color()
+    local mp_format = string.format('[' .. get_mp() .. ']{' .. mp_format_string .. '}')
+
+    local tp_format_string = options.text_typeface .. ' ' .. tostring(options.text_size) .. 'px bold ' .. get_tp_color()
+    local tp_format = string.format('[' .. get_tp() .. ']{' .. tp_format_string .. '}')
+
+    local debuffs_format_string = options.text_typeface .. ' ' .. tostring(options.text_size) .. 'px bold ' .. 'white'
+    local debuffs_format = string.format('[' .. party_data[1].debuffs .. ']{' .. debuffs_format_string .. '}')
+
+    return name_job_hp_format .. mp_format .. tp_format .. debuffs_format
+end
+
+local function check_target_for_row_highlighting()
+    local entity = target.t or target.st
+    for i = 1, 6 do
+        if entity and party[i] then
+            if entity.name == party[i].name then
+                party_window_states[i].color = ui.color.rgb(30, 30, 30)
+            else
+                party_window_states[i].color = ui.color.black
+            end
+        else
+            party_window_states[i].color = ui.color.black
+        end
+    end
+end
+
+-- Player zone out packet
+packets.incoming[0x00B]:register(function()
+    zone_out_time = os.clock()
+    zoning = true
+end)
+
+-- 0x0DF appears to change when player hp/mp/tp changes
+packets.incoming[0x0DF]:register(function()
+    update_party_name_hp_mp_tp_strings('both')
+    update_party_debuff_strings('both')
+end)
+
+-- This seems to change when buffs on self change, also when party members zone
+packets.incoming[0x037]:register(function()
+    update_party_name_hp_mp_tp_strings('both')
+    update_party_debuff_strings('both')
+end)
+
+-- Party member updates, joining/leaving party etc
+packets.incoming[0x0DD]:register(function()
+    update_party_name_hp_mp_tp_strings('both')
+    update_party_debuff_strings('both')
+end)
+
+-- Party buffs update packet
+packets.incoming[0x076]:register(function()
+    update_party_debuff_strings('party_only')
+end)
+
+ui.display(function()
+    if zoning and os.clock() > zone_out_time + zone_in_delay then
+        zoning = false
+    end
```

Comment:
What purpose does the zone timeout serve? You can probably avoid it entirely by looking for certain packets.

Reply:
The timeout here is used to hide the UI until party data has been correctly populated into the display after a player zones. I tried to implement this with the incoming packet 0x00A but I found it would always show the UI before party information was correctly populating the display. What other packets could possibly be used do you think?

Reply:
Do you have any insight as to what the flags in the 0x0DD packet are?
```lua
types.incoming[0x0DD] = struct({
    player_id           = {0x00, entity},
    hp                  = {0x04, uint32},
    mp                  = {0x08, uint32},
    tp                  = {0x0C, uint32},
    flags               = {0x10, uint16},
    player_index        = {0x14, entity_index},
    hp_percent          = {0x19, percent},
    mp_percent          = {0x1A, percent},
    zone_id             = {0x1C, zone},
    main_job_id         = {0x1E, job},
    main_job_level      = {0x1F, uint8},
    sub_job_id          = {0x20, job},
    sub_job_level       = {0x21, uint8},
    player_name         = {0x22, pc_name},
})
```
After doing a little bit of testing my best guess at the moment is the most significant digit indicates the alliance leader, the next digit indicates a party leader, and the last two digits indicate the party number, where the party number is 0 if a party is not part of an alliance, 1 if it is the alliance leader party, 2 and 3 would be the other not alliance lead parties I guess.

I think I do need to figure out how to only call the inside of the following once, when this packet gets spammed once for each party member in party/alliance I think:
```lua
packets.incoming[0x0DD]:register(function()
    update_party_name_hp_mp_tp_strings('both')
    update_party_status_strings('both')
end)
```
Not sure how to do that at the moment using these flags, any ideas?

Reply:
It might be better to use 0x0C8.

Reply:
That would actually help a lot. That way you know exactly which `0x0DD` you can expect, and once you receive all of them you can show the alliance list.